File: Compiler/Abi/Manifest.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using Compiler.Llvm;
using Compiler.Semantics;
using Compiler.Syntax;
using Type = Compiler.Semantics.Type;

namespace Compiler.Abi;

// The contract between a compiled pod and its host: the depth limit the code enforces, the
// errors the code can fail with, every struct with the offsets of its fields, and every function.
// An external function comes with the symbol of its entry and the layout of its argument and
// result blocks; an internal one is listed by name only, for the traces. The host reads the
// manifest once when loading the pod, then only ever passes memory. An argument marked ref is
// one the function writes back into the block it was given; a function marked fails ends its
// results with its error. Functions are listed in the order of their numbers, which the frames
// of a trace refer to, and Source names the file for the lines of those frames. State is the
// block of the variables the pod keeps between calls, which the host lays out in the instance.
public sealed record Manifest(string Source, int MaxDepth, List<ErrorName> Errors, Dictionary<string, Block> Structs, Block State, List<Function> Functions)
{
    private static readonly JsonSerializerOptions jsonOptions = new() {
        WriteIndented = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };


    // Every function gets an entry the host can call, named after it.
    public static string SymbolOf(string function)
    {
        return $"bc_{function}";
    }


    public static Manifest Build(ProgramNode program, TypeChecker checker, string source)
    {
        var structs = new Dictionary<string, Block>();
        foreach (var declaration in program.Structs) {
            var type = checker.Structs[declaration];
            structs[type.Name] = Block.Of(type.Fields.Select(f => ((string?)f.Name, f.Type, false)));
        }

        var errors = checker.Errors.Select(e => new ErrorName(e.Name, e.Code)).ToList();

        var functions = new List<Function>();
        foreach (var declaration in program.Functions) {
            var symbol = checker.Functions[declaration];

            if (!symbol.IsExternal) {
                functions.Add(new Function(symbol.Name, null, null, null, null, null));
                continue;
            }

            var parameters = symbol.Parameters.Select(p => ((string?)p.Name, p.Type, p.IsRef));
            var results = IrType.Results(symbol).Select(t => ((string?)null, t, false));
            functions.Add(new Function(symbol.Name, true, SymbolOf(symbol.Name), symbol.CanFail ? true : null, Block.Of(parameters), Block.Of(results)));
        }

        var state = Block.Of(checker.States.Select(s => ((string?)s.Name, s.Type, false)));
        return new Manifest(source, Protocol.MaxDepth, errors, structs, state, functions);
    }


    public string ToJson()
    {
        return JsonSerializer.Serialize(this, jsonOptions);
    }
}


public sealed record ErrorName(string Name, int Code);


// External and Fails are only ever written when true; an internal function has only its name.
public sealed record Function(string Name, bool? External, string? Symbol, bool? Fails, Block? Args, Block? Results);


// Values laid out one after the other: a struct, or the arguments or results of a function.
// Results have no names.
public sealed record Block(int Size, List<Field> Fields)
{
    public static Block Of(IEnumerable<(string? Name, Type Type, bool IsRef)> members)
    {
        var list = members.ToList();
        var (offsets, size) = Layout.Place(list.Select(m => m.Type));
        var fields = list.Select((m, i) => new Field(m.Name, m.Type.ToString(), offsets[i], m.IsRef ? true : null)).ToList();
        return new Block(size, fields);
    }
}


// Ref is only ever written when true, on the arguments a function takes by ref.
public sealed record Field(string? Name, string Type, int Offset, bool? Ref);

File: Runtime/manifest.go
package bluecode

import (
	"encoding/json"
	"fmt"
	"strings"
)

// Manifest is what the compiler emits next to a pod: the depth limit its code enforces, the
// errors its code can fail with, the memory layout of every struct, and for every function the
// symbol of its entry and the layout of its argument and result blocks.
// Offsets follow the natural alignment rule that LLVM, C and Go share, so a Go struct with
// the same fields in the same order is laid out identically and can be passed as is.
// Functions are in the order of their numbers, which the frames of a trace refer to.
type Manifest struct {
	Source    string           `json:"source"`   // the file the pod was compiled from, for the lines of a trace
	MaxDepth  int              `json:"maxDepth"` // how deep the code lets calls nest before giving up
	Errors    []ErrorName      `json:"errors"`
	Structs   map[string]Block `json:"structs"`
	State     Block            `json:"state"` // the variables the pod keeps between calls, as one block
	Functions []Signature      `json:"functions"`
}

// ErrorName is an error the pod declares, with the number its code fails with.
type ErrorName struct {
	Name string `json:"name"`
	Code int64  `json:"code"`
}

// Signature describes one function. Only an External one can be called: it has the Symbol of
// its entry, Args, the block its parameters form, and Results, the block its return values are
// written to. A function that Fails ends its results with an "error" field holding the number
// of the error it failed with, zero on success. An internal function is listed by name only,
// so that the frames of a trace can be read.
type Signature struct {
	Name     string `json:"name"`
	External bool   `json:"external"`
	Symbol   string `json:"symbol"`
	Fails    bool   `json:"fails"`
	Args     Block  `json:"args"`
	Results  Block  `json:"results"`
}

// Block is a run of values laid out one after the other, padding included in Size.
type Block struct {
	Size   int     `json:"size"`
	Fields []Field `json:"fields"`
}

// Field is one value of a block. Type is "int", "uint", "bool", "error" or the name of a
// struct. Results have no name. Ref marks an argument the function works on in place: after
// the call, the caller's block holds what the function left there.
type Field struct {
	Name   string `json:"name"`
	Type   string `json:"type"`
	Offset int    `json:"offset"`
	Ref    bool   `json:"ref"`
}

func parseManifest(data []byte) (*Manifest, error) {
	var manifest Manifest
	if err := json.Unmarshal(data, &manifest); err != nil {
		return nil, fmt.Errorf("manifest: %w", err)
	}

	return &manifest, nil
}

// Canonical checks that data is the one valid representation of a block: exactly its size,
// every bool 0 or 1, every padding byte 0. A node runs this on the bytes of a transaction
// before handing them to a pod, so that a value has a single encoding and a single hash.
func (m *Manifest) Canonical(block Block, data []byte) error {
	if len(data) != block.Size {
		return fmt.Errorf("block is %d bytes, expected %d", len(data), block.Size)
	}

	covered := make([]bool, len(data))
	if err := m.walk(block, data, 0, covered); err != nil {
		return err
	}

	for i, isCovered := range covered {
		if !isCovered && data[i] != 0 {
			return fmt.Errorf("padding byte %d is not zero", i)
		}
	}

	return nil
}

func (m *Manifest) walk(block Block, data []byte, base int, covered []bool) error {
	for _, field := range block.Fields {
		at := base + field.Offset

		if strings.HasPrefix(field.Type, "own ") {
			return fmt.Errorf("field %q owns memory of the pod: it never crosses to the host", field.Name)
		}

		switch field.Type {
		case "int", "uint", "error":
			for i := at; i < at+8; i++ {
				covered[i] = true
			}

		case "bool":
			if data[at] > 1 {
				return fmt.Errorf("byte %d is %d, not a bool", at, data[at])
			}
			covered[at] = true

		default:
			inner, ok := m.Structs[field.Type]
			if !ok {
				return fmt.Errorf("unknown type %q", field.Type)
			}
			if err := m.walk(inner, data, at, covered); err != nil {
				return err
			}
		}
	}

	return nil
}
